src/FastRouteRouter.php

```diff
@@ -277,41 +261,50 @@ public function generateUri($name, array $substitutions = [], array $options = [
         }
 
         $route   = $this->routes[$name];
-        $path    = $route->getPath();
         $options = ArrayUtils::merge($route->getOptions(), $options);
 
         if (! empty($options['defaults'])) {
             $substitutions = array_merge($options['defaults'], $substitutions);
         }
 
-        foreach ($substitutions as $key => $value) {
-            $pattern = sprintf(
-                '~%s~x',
-                sprintf(self::VARIABLE_REGEX, preg_quote($key))
-            );
-            $path = preg_replace($pattern, $value, $path);
-        }
+        $routeParser = new RouteParser();
+        $routes = $routeParser->parse($route->getPath());
+
+        // One route pattern can correspond to multiple routes if it has optional parts
+        foreach ($routes as $r) {
+            $path = '';
+            $paramIdx = 0;
+            foreach ($r as $part) {
+                // Fixed segment in the route
+                if (is_string($part)) {
+                    $path .= $part;
+                    continue;
+                }
+
+                // Placeholder in the route
+                if ($paramIdx === count($substitutions)) {
+                    throw new LogicException('Not enough parameters given');
+                }
 
-        // 1. remove optional segments' ending delimiters
-        //    and remove leftover regex char classes like `:[` and `:prefix-[` (issue #18)
-        // 2. split path into an array of optional segments and remove those
-        //    containing unsubstituted parameters starting from the last segment
-        $path = preg_replace('/\]|:.*\[/', '', $path);
-        $segs = array_reverse(explode('[', $path));
-        foreach ($segs as $n => $seg) {
-            if (strpos($seg, '{') !== false) {
-                if (isset($segs[$n - 1])) {
+                if (! isset($substitutions[$part[0]])) {
                     throw new Exception\InvalidArgumentException(
                         'Optional segments with unsubstituted parameters cannot '
```

Comment:
I think we should use a message similar to what @michaelmoussa proposed earlier: `Expected parameter values for [id,foo,bar], but received [id,foo,not_bar]`. Again, this would demonstrate to the developer what occurred so they have a reasonable chance of correcting the situation in the calling code.